### ISSUE TITLE
Storage vs memory variable

### ISSUE THREAD

**Issue by aleybovich** [saurfang/ipfs-multihash-on-solidity#1]
In line 73: `Multihash storage entry = entries[_address];` you should really use `memory` instead of 'storage' - you are not storing `entry` variable after you assign it on this line. As far as I remember, using `storage` keyword on local variables may cause undefined/unexpected behavior as well.

**Comment by saurfang** [comment]
Thank you for your suggestion. Can you link any reference on the claim `using storage keyword on local variables may cause undefined/unexpected behavior`? 

As far as I understand, `Multihash storage entry = entries[_address];` is dereferencing from a pre-allocated storage, namely `entries`. `entry` is merely a storage pointer to the preallocated storage section for the contract state variable.

I looked at this FAQ:
http://solidity.readthedocs.io/en/v0.4.21/frequently-asked-questions.html#what-is-the-memory-keyword-what-does-it-do
They gave the example that `uint[] x;` would reference the storage slot `0` but `uint[] x = data;` would reference to the preallocated `uint[] data;` contract variable in the storage. I take that to imply that dereferencing contract variable to a local variable merely creates a pointer but not allocating additional variables.

I think what you are suggesting is whether `Multihash storage entry = entries[_address];` gets allocated to a new slot on the storage. In fact, `Multihash entry = entries[_address];` is also legal and according to the following section, it defaults to `storage` location, i.e., equivalent to what we have currently.
https://solidity.readthedocs.io/en/v0.4.21/types.html#data-location

The code example here provides an example that is similar to ours, where there is a mapping of address to struct, and `Campaign storage c = campaigns[campaignID];` is retrieved and modified. I take that as an evidence that in fact no new storage is allocated for `c` otherwise the modification would not persist back to `campaigns`. In this function, `c` merely points to a section of storage within `campaigns`.
https://solidity.readthedocs.io/en/v0.4.21/types.html#structs

My understanding is that `memory` should only be used for local variable that is newly created but do not persist in the contract. For example, in the `setEntry` function, we create a memory pointer `entry` because its content is copied to `entries` in the next line and `entry` itself should not persist.
```
Multihash memory entry = Multihash(_digest, _hashFunction, _size);
entries[msg.sender] = entry;
```
In this case, it isn't even possible to compile `Multihash storage entry = Multihash(_digest, _hashFunction, _size);` because RHS creates a memory pointer.


To recap, I think if we change to `Multihash memory entry = entries[_address];`, it might allocate a new variable on memory instead and results in higher gas cost but is not any more correct. I made a new function [`getEntryMemory`](
https://github.com/saurfang/ipfs-multihash-on-solidity/commit/7d649635aa109a322ea9a35e8d051b632807844e) and at least truffle's `estimateGas` seems to suggest it would cost `24167` instead of `23938` before.
https://travis-ci.org/saurfang/ipfs-multihash-on-solidity/builds/354887877#L1273

**Comment by aleybovich** [comment]
@saurfang I didn't find anything about unexpected behavior, instead, I found examples of why in some cases you do want to use `storage` when declaring a variable in a function - that is when you want to `copy` by reference and further be able to modify storage if modifying that variable.
Actually, now that I've researched more, I *think* I understand why you get less gas usage when using `storage` - by declaring your variable `storage` you are not taking up any additional memory, you are simply referencing the stored value; if you declare that variable `entry` as `memory` it will take up additional memory in the VM thus extra gas cost.

So yes, you are right :) And I've learned something today!

**Comment by saurfang** [comment]
Me too! Thank you for asking this great question.
